File: algorithms/java/Graph/src/com/mahbub/algorithm/SecondBestSpanningTree_Demo.java
package com.mahbub.algorithm;

import org.omg.CORBA.INTERNAL;

import java.io.FileInputStream;
import java.io.FileNotFoundException;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Scanner;

public class SecondBestSpanningTree_Demo {

    static int NODES, EDGES;
    static List<Edge> edges;
    static List<Integer> spanEdges;
    static int mstCost;

    // -------------------- Disjoint Set Structure --------------------------------------
    static int[] dset;
    static void initSet(int N) {
        dset = new int[N+1];
        for (int i = 0; i <= N; i++) dset[i] = i;
    }
    static int findSet(int u) {
        return dset[u] == u ? u : (dset[u] = findSet(dset[u]));
    }
    static void union(int u, int v) {
        dset[findSet(u)] = findSet(v);
    }
    // ----------------------------------------------------------------------------------

    static class Edge implements Comparable<Edge> {
        int u, v, c;
        Edge(int u, int v, int c) {
            this.u = u;
            this.v = v;
            this.c = c;
        }

        @Override
        public int compareTo(Edge o) {
            return this.c - o.c;
        }
    }

    static int kruskal(int excludedEdgeIndex) {
        int countSpanEdges = 0;
        mstCost = 0;

        Collections.sort(edges);
        initSet(NODES);
        spanEdges = new ArrayList<>();

        for (int i = 0; i < edges.size(); i++) {
            if (i == excludedEdgeIndex) continue;
            if (countSpanEdges == NODES - 1) break;
            Edge e = edges.get(i);
            if (findSet(e.u) != findSet(e.v)) {
                spanEdges.add(i);
                union(edges.get(i).u, e.v);
                mstCost += e.c;
                countSpanEdges++;
            }
        }
        return mstCost;
    }

    static int secondBestMST() {
        kruskal(-1);
        List<Integer> spanEdgeIndexes = new ArrayList<>(spanEdges);

        int secondBestCost = Integer.MAX_VALUE;
        for (int i = 0; i < spanEdgeIndexes.size(); i++) {
            secondBestCost = Math.min(secondBestCost, kruskal(spanEdgeIndexes.get(i)));
        }

        return secondBestCost;
    }

    public static void main(String[] args) {
        Scanner scanner = inputFromFile();

        while (scanner.hasNext()) {
            NODES = scanner.nextInt();
            EDGES = scanner.nextInt();

            edges = new ArrayList<>();

            for (int i = 0; i < EDGES; i++) {
                int u = scanner.nextInt();
                int v = scanner.nextInt();
                int c = scanner.nextInt();

                edges.add(new Edge(u, v, c));
            }

            int secondBestMSTCost = secondBestMST();

            System.out.println("Second best MST Cost: " + secondBestMSTCost);
            for (int i = 0; i < spanEdges.size(); i++) {
                Edge e = edges.get(spanEdges.get(i));
                System.out.println(e.u + " " + e.v + " " + e.c);
            }
        }
    }

    static Scanner inputFromFile() {
        try { return new Scanner(new FileInputStream("input.txt")); }
        catch (FileNotFoundException e) { e.printStackTrace(); }
        return null;
    }

    static Scanner inputFromSystem() {
        return new Scanner(System.in);
    }

}

/**
 INPUT:
 5 8
 0 1 9
 0 2 75
 1 2 95
 1 4 42
 1 3 19
 3 4 31
 2 3 51
 2 4 66

 OUTPUT:
 Second best MST Cost: 121
 0 1 9
 1 3 19
 3 4 31
 2 4 66
**/

File: algorithms/kotlin/Graph/SecondBestSpanningTree_Demo.kt
package com.mahbub.algorithm.graph

import java.io.FileInputStream
import java.io.FileNotFoundException
import java.util.*
import javax.print.attribute.IntegerSyntax
import kotlin.collections.ArrayList
import kotlin.math.min

class SecondBestSpanningTree_Demo {
    class Edge(u : Int, v : Int, c : Int) : Comparable<Edge> {
        var u = u
        var v = v
        var c = c
        override fun compareTo(o: Edge): Int {
            return this.c - o.c
        }
    }

    // -------------------- Disjoint Set Structure --------------------------------------
    private var dset: IntArray = IntArray(0)

    fun initSet(N: Int) {
        dset = IntArray(N + 1)
        for (i in 0..N) dset[i] = i
    }

    fun findSet(u: Int): Int {
        var ret : Int
        if (dset[u] == u)
            ret = u
        else {
            dset[u] = findSet(dset[u])
            ret = dset[u]
        }
        return ret
    }

    fun union(u: Int, v: Int) {
        dset[findSet(u)] = findSet(v)
    }
    // ----------------------------------------------------------------------------------


    var NODES : Int = 0
    var EDGES : Int = 0
    var edges : MutableList<Edge> = ArrayList()
    var spanEdges : MutableList<Int> = ArrayList()
    var mstCost : Int = 0

    fun kruskal(excludedEdgeIndex : Int) : Int {
        var countSpanEdges = 0
        mstCost = 0

        edges.sort()
        initSet(NODES)
        spanEdges = ArrayList()

        for (i in 0 until edges.size) {
            if (i == excludedEdgeIndex) continue
            if (countSpanEdges == NODES-1) break
            var e = edges[i]
            if (findSet(e.u) != findSet(e.v)) {
                spanEdges.add(i)
                union(edges[i].u, e.v)
                mstCost += e.c
                countSpanEdges++
            }
        }
        return mstCost
    }

    fun secondBestMST() : Int {
        kruskal(-1)
        var spanEdgeIndexes = ArrayList<Int>(spanEdges)

        var secondBestCost = Integer.MAX_VALUE
        for (i in spanEdgeIndexes) {
            secondBestCost = min(secondBestCost, kruskal(i))
        }
        return secondBestCost
    }

    fun fakeMain() {
        var scanner = inputFromFile()

        while (scanner!!.hasNext()) {
            NODES = scanner.nextInt()
            EDGES = scanner.nextInt()

            edges = ArrayList()

            for (i in 0 until EDGES) {
                var u = scanner.nextInt()
                var v = scanner.nextInt()
                var c = scanner.nextInt()

                edges.add(Edge(u, v, c))
            }

            var secondBestMSTCost = secondBestMST()

            println("Second best MST Cost: $secondBestMSTCost")
            for (i in 0 until spanEdges.size) {
                var e = edges[spanEdges[i]]
                println("${e.u} ${e.v} ${e.c}")
            }
        }
    }
}

fun main(args : Array<String>) {
    SecondBestSpanningTree_Demo().fakeMain()
}

private fun inputFromFile() : Scanner? {
    return try { Scanner(FileInputStream("input.txt")) }
    catch (e : FileNotFoundException) { null }
}

private fun inputFromSystem() : Scanner? {
    return Scanner(System.`in`)
}